# Profitable app profiles for the App store and Google play markets.

* The company that builds free downloadable Android and iOS apps accessible on google play and the App store. Revenues for the company are generated from ads displayed on its apps and would like more of their app users to engage with the adds displayed. 

* The goal from the analysis is to assist app developers at the company to make data driven decisions in determinig what apps to design that could attrack more users on both Google play and the App store. Datasets from both Google play and the App store were collected for the analysis.  

In [1]:
# reading the data sets into lists
Apple_store_data = open('AppleStore.csv')
Google_store_data = open('googleplaystore.csv')
from csv import reader
Apple = reader(Apple_store_data)
Google = reader(Google_store_data)
Apple_data = list(Apple)
Google_data = list(Google)

In [2]:
# exploring the data sets
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))
        
# first few rows of each data set
explore_data(Apple_data, 1, 3)
print('\n')
explore_data(Google_data, 1, 3)

# number of rows and colums of each data set
Apple_no_header = Apple_data[1:]
explore_data(Apple_no_header, 0, 3, True)
print('\n')
Google_no_header = Google_data[1:]
explore_data(Google_no_header, 0, 3, True)


        


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']




['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0',

In [3]:
# column names
Apple_header = Apple_data[0]
Google_header = Google_data[0]
print(Apple_header)
print('\n')
print(Google_header)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


The columns of the Apple iOS data that could be helpful for the analysis are track_name, currency, price, rating count_tot, rating_count_ver, cont_rating and prime_genre. The documentation for the columns can be found here [Link](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/home)

The columns for the Google play data that could be useful for the analysis are Category, Rating, reviews, Installs, Type, Price and Genres. The documantation for the columns can be found here [Link](https://www.kaggle.com/lava18/google-play-store-apps/home#googleplaystore.csv)

In [4]:
# Finding the error row
Google_data[0]



['App',
 'Category',
 'Rating',
 'Reviews',
 'Size',
 'Installs',
 'Type',
 'Price',
 'Content Rating',
 'Genres',
 'Last Updated',
 'Current Ver',
 'Android Ver']

In [5]:
Google_data[10473]

['Life Made WI-Fi Touchscreen Photo Frame',
 '1.9',
 '19',
 '3.0M',
 '1,000+',
 'Free',
 '0',
 'Everyone',
 '',
 'February 11, 2018',
 '1.0.19',
 '4.0 and up']

The Rating column for row 10472 has the entry 19 as a reult of the rest of the data being shifted to the right. The maximum rating is 5 which makes it incorrect.

In [6]:
# Deleting the row with an error
del Google_data[10473]


In [7]:
# With the header row
print(len(Google_data))

# Without the header row
print(len(Google_data[1:]))

10841
10840


After exploring the Google play data set extensively there are instances that certain apps have duplicate entries for instance Instagram below.

In [8]:
for app in Google_data[1:]:
    name = app[0]
    if name == 'Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


In [9]:
# counting the number of duplicate apps
duplicate_apps = []
unique_apps = []

for app in Google_data[1:]:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
    
print('Number of duplicate apps:', len(duplicate_apps))
print('\n')
print('Examples of duplicate apps:', duplicate_apps[:15])

Number of duplicate apps: 1181


Examples of duplicate apps: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


Certain duplicate apps can be removed randomly however from observing the rows of a duplicate app for instance Instagram the number of reviews collected for the app is unique as a result of being collected at different times. This will serve as the criteria for retaining a row and discarding duplicates. The highest number of reviews for given set of duplicate rows will ensure that the row with the most reliable rating is kept.   

In [10]:
# removing duplicate rows using its maximum number of reviews

reviews_max = {} 

for row in Google_data[1:]:
    name = row[0]
    n_reviews = float(row[3])
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
    elif name not in reviews_max:
        reviews_max[name] = n_reviews


In [11]:
# exploring the dictorany to verify the expected number of unique apps

print('Number of apps without duplicates: ', len(Google_data[1:]) - 1181)
print('Number of unique apps stored in maximum reviews dictionary: ', len(reviews_max))

Number of apps without duplicates:  9659
Number of unique apps stored in maximum reviews dictionary:  9659


In [12]:
# removing duplicate rows and storing the data in lists

android_clean = []
already_added = []

for row in Google_data[1:]:
    name = row[0]
    n_reviews = float(row[3])
    if (n_reviews == reviews_max[name]) and (name not in already_added):
        android_clean.append(row)
        already_added.append(name)



In [13]:
# exploring the stored non duplicate data set
explore_data(android_clean, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9659
Number of columns: 13


The android_clean data list has 9659 rows as expected

In [14]:
# detecting English character strings
def eng_string(input_string):
    for character in input_string:
        if ord(character) > 127:
            return False
    return True
        

In [15]:
print(eng_string('Instagram'))
print(eng_string('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(eng_string('Docs To Go™ Free Office Suite'))
print(eng_string('Instachat 😜'))

True
False
False
False


In [16]:
# removing non-Engish app names from the data sets

def eng_string2(input_string):
    char_count = 0
    for character in input_string:
        if ord(character) > 127:
            char_count += 1
            
    if char_count > 3:
        return False
    else:
        return True
    
print(eng_string2('Docs To Go™ Free Office Suite'))
print(eng_string2('Instachat 😜')) 
print(eng_string2('爱奇艺PPS -《欢乐颂2》电视剧热播'))

True
True
False


In [17]:
# storing only English apps in two lists

eng_apps_Google = []
eng_apps_Apple = []

for row in android_clean:
    app = row[0]
    if eng_string2(app):
        eng_apps_Google.append(row)

for row in Apple_data[1:]:
    app = row[1]
    if eng_string2(app):
        eng_apps_Apple.append(row)
        
# exploring number of rows left for each data set
explore_data(eng_apps_Google, 0, 3, True)
print('\n')
explore_data(eng_apps_Apple, 0, 3, True)
    

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns: 13


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 

We can see that we're left with 9614 Android apps and 6183 iOS apps.

In [18]:
# finding the index for price in both data sets
print(Google_data[0])
print(Apple_data[0])

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


In [19]:
# storing free apps in a list

Google_free_apps = []
Apple_free_apps = []

for row in eng_apps_Google:
    price = row[7]
    if price == '0':
        Google_free_apps.append(row)
        
for row in eng_apps_Apple:
    price = row[4]
    if price == '0.0':
        Apple_free_apps.append(row)
        
print('Number of free Google apps: ', len(Google_free_apps))
print('Number of free Apple apps: ', len(Apple_free_apps))

Number of free Google apps:  8864
Number of free Apple apps:  3222


There are 8864 Google android apps and 3222 Apple iOS apps left for analysis

* Since the company deals with free downloadable apps and generates revenue from ads run on those apps it is beneficial to find an app profile that fits the criteria of being popularly used on both Google Play and the Apple Store. The validation stratergy for an app idea comprises of building a minimal android version of the app and testing it on Google Play for responses from users. If the responses are good we develop the app further and if it has made profits after six months an iOS version is built for the App Store. 

* For the Google android data sets the columns that can be used to find the most common genres in each market are Category, Rating, Reviews, Installs, Price and Genres. For Apple iOS the columns are price, rating_count_tot, user_rating, cont_rating and prime_genre. 






In [20]:
# creating a frequency table in decending order

def freq_table(dataset, index): # percent frequency table
    table = { }
    count = 0
    for row in dataset:
        data = row[index]
        if data in table:
            table[data] += 1
        else:
            table[data] = 1
        count+= 1 
    #print(count)
    #print(len(table))
    
    percentage_table = { }
    for key in table:
        #freq_table[key] *= 100 / len(freq_table) 
        percentage = (table[key] / count) * 100
        percentage_table[key] = percentage
        
    #print(len(percentage_table))    
    return percentage_table
            
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])
display_table(Apple_free_apps, -5)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


The most common genre is Games with 58.2% followed by Entertainment by 7.9%. More than half of the genre of apps in the App store belongs to Games while the other 41.8% belongs to other genres each of which were below 10%. The general impression is that most apps in the App store are used for fun (games, entertainment, photo and video, social networking, sports, music, etc.). Recomendations of an app profile based on the frequency table alone will be inaccurate. consideration has to be given to rating count as their demand might vary. 

In [21]:
# frequency table for Category column
display_table(Google_free_apps, 1)

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

In [22]:
# frequency table for Genre column
display_table(Google_free_apps, -4)

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

The most common genres for the columns are family and tools. The next most common app for the Category column games is about half of the number of family apps 18.9% on Google play at 9.7%. The frequencies for the genres of apps for Google Store are evenly distributed compared to that of the App Store. Also, App store have more genre of apps that are for fun compared to Google play apps that have an even distribution of apps for practical use and fun. Recommendations for app development cannot be done without further expolring other factors that indicate the demand patterns of users like the number of users or number of downloads and ratings of the app.    

In [23]:
# finding the most popular apps on Apple Store by genre

genres_app_store = freq_table(Apple_free_apps, -5)

for genre in genres_app_store:
    total = 0
    len_genre = 0
    for row in Apple_free_apps:
        genre_app = row[-5]
        if genre_app == genre:
            n_ratings = float(row[5])
            total += n_ratings
            len_genre += 1
    avg_n_ratings =  total / len_genre
    print(genre, ':', avg_n_ratings)
    



Finance : 31467.944444444445
News : 21248.023255813954
Reference : 74942.11111111111
Navigation : 86090.33333333333
Sports : 23008.898550724636
Music : 57326.530303030304
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Travel : 28243.8
Productivity : 21028.410714285714
Weather : 52279.892857142855
Utilities : 18684.456790123455
Catalogs : 4004.0
Education : 7003.983050847458
Health & Fitness : 23298.015384615384
Shopping : 26919.690476190477
Food & Drink : 33333.92307692308
Social Networking : 71548.34905660378
Book : 39758.5
Photo & Video : 28441.54375
Medical : 612.0
Games : 22788.6696905016
Business : 7491.117647058823


Navigation apps have the largest number of user ratings followed by Social networking apps. Since the App Store predominantly has more fun apps perhaps the company can design Social networking apps. Reference apps can also be designed as well since they have the second highest number of reviews. 

A social networking app for instance Facebook can be designed with the capability of peharps having fun features like a music widget or video chat capabilities or even widgets for navigation or references like for news articles or an encyclopedia to get the maximum engagement with the app that the company needs for its free downloadable app.

In [30]:
# finding the average number of installs per genre for Google play

freq_google_data = freq_table(Google_free_apps, 1)

for category in freq_google_data:
    total = 0
    len_category = 0
    for app in Google_free_apps:
        category_app = app[1]
        if category_app == category:
            n_installs = app[5]
            n_installs = n_installs.replace(',', '')
            n_installs = n_installs.replace('+', '')
            total += float(n_installs)
            len_category += 1
    ave_n_installs = total / len_category
    print(category, ':', ave_n_installs)
    


NEWS_AND_MAGAZINES : 9549178.467741935
DATING : 854028.8303030303
MEDICAL : 120550.61980830671
LIBRARIES_AND_DEMO : 638503.734939759
EVENTS : 253542.22222222222
VIDEO_PLAYERS : 24727872.452830188
BUSINESS : 1712290.1474201474
EDUCATION : 1833495.145631068
WEATHER : 5074486.197183099
COMMUNICATION : 38456119.167247385
HOUSE_AND_HOME : 1331540.5616438356
TRAVEL_AND_LOCAL : 13984077.710144928
HEALTH_AND_FITNESS : 4188821.9853479853
FINANCE : 1387692.475609756
PARENTING : 542603.6206896552
TOOLS : 10801391.298666667
PERSONALIZATION : 5201482.6122448975
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
AUTO_AND_VEHICLES : 647317.8170731707
ART_AND_DESIGN : 1986335.0877192982
GAME : 15588015.603248259
PRODUCTIVITY : 16787331.344927534
PHOTOGRAPHY : 17840110.40229885
SOCIAL : 23253652.127118643
LIFESTYLE : 1437816.2687861272
SPORTS : 3638640.1428571427
SHOPPING : 7036877.311557789
COMICS : 817657.2727272727
ENTERTAINMENT : 11640705.88235294
FOOD_AND_DRINK : 1924897.736363636

Communication apps for Google play have the highest number of installs with popular apps like Facebook, Skype, Whatsapp and messanger. Social apps are the next popular apps followed by Entertainment apps. Since fun apps more that function apps are popular on both Google play and the App Store an app recomendation for both platforms can be a social networking site that has functionality of engaging users by having communication capabilities like video chat and instant messaging.

Another option might be designing a functional app like a navigational app or an educational app that the users can engage with. A free language app or sports app can be engaging enough to run the company's ads. 